### PR TITLE
feat: teach PyVortex to read from object storage

### PR DESCRIPTION
**Pull request by danking**
Credentials are sourced from the environment or the metadata server as appropriate. Files, R2, and plain-old-HTTP(S) all work. In principle, it also supports GCS, S3, and Azure; however, I have not tested them.

```python
import vortex

a = vortex.io.read(url='https://gist.github.com/danking/829d802b6d9ed9f30576bf628a796f4d/raw/47bb3d4359da43f2e29c177cae52391438311119/testfile.vortex')
a.to_arrow_table()

c = vortex.io.read(url='https://01e9655179bbec953276890b183039bc.r2.cloudflarestorage.com/vortex-test/yellow-tripdata-2023-11.vortex')
c.to_arrow_table()

d = vortex.io.read(url='file:/Users/danielking/projects/vortex/bench-vortex/data/PBI/Food/vortex/Food_1.vortex')
d.to_arrow_table()

e = vortex.io.read(file='/Users/danielking/projects/vortex/bench-vortex/data/PBI/Food/vortex/Food_1.vortex')
e.to_arrow_table()
```

This PR is a bit hairy:
1. I added a feature-flagged `impl From<VortexError> for PyErr` which simplified a lot of use-cases, particularly in my new code where all the `map(PyVortexError::map_err)` really obscured the meaning.
2. I wrote my own `better_parse_url` because Arrow/object_store's doesn't include the environment by default. If AWS_SECRET_ACCESS_KEY is in my environment I expect things to Just Work.
3. I cannot figure out how to abstract over `VortexReadAt`. The trait isn't object-safe. Trying to stuff two options into an enum and making that enum `VortexReadAt` ran into issues with the `impl Future` return type.
4. I make the user be specific about whether they have a URL versus a path because I don't want to triage bug reports about files with a literal `%20` in their name.

### CODE REVIEW

File: pyvortex/src/dataset.rs [hunk 1]
```diff
@@ -102,27 +101,31 @@ impl PyDataset {
...
 }
 
 #[pyfunction]
-pub fn dataset(fname: &Bound<'_, PyString>) -> PyResult<PyDataset> {
-    TOKIO_RUNTIME
-        .block_on(PyDataset::new(fname.to_str()?))
-        .map_err(PyVortexError::map_err)
+#[pyo3(signature = (*, file = None, url = None))]
+pub fn dataset(
+    file: Option<&Bound<PyString>>,
```
Comment (gatesn) [inline pyvortex/src/dataset.rs]: Having two mutually exclusive arguments is a bit odd in Python. Either it should be two functions, or a single argument that is switched on internally, e.g. `try { url = urlparse(url) } except { file = url }`

Comment (danking) [reply pyvortex/src/dataset.rs]: Most paths will parse successfully. I could bail out if the scheme is empty; but, that gives heebie-jeebies. I don't think I understand all the path syntaxes enough to be certain I'd correctly handle every one. Other than Hail, the only example I know which uses mutually exclusive kwargs is `requests` with `data` vs `json`.

I'm happy to go with `dataset_from_url`, `dataset_from_path`, `read_url` and `read_path`, if that passes muster.

Comment (danking) [reply pyvortex/src/dataset.rs]: @gatesn I switch to the separated to separate functions and merged main; ready for another look.

**Comment by danking** [comment]
Bump for re-review @gatesn @robert3005

**Comment by robert3005** [comment]
@danking my suggestion was to make VortexReadAt an object safe trait. Otherwise users are required to use our implementations of the readers

**Comment by danking** [comment]
Alright, VortexReadAt is object-safe and PyDataset holds an `Arc<dyn VortexReadAt>`.

**Comment by danking** [comment]
See https://github.com/spiraldb/vortex/pull/1151

**Pull request closed without merging**